File: 2016/06.cs
namespace _2016;

public abstract class _06
{
    private static readonly string[] Data;

    static _06() => Data = Task.Run(() => Utils.GetInputData(6)).Result
        .Split('\n', StringSplitOptions.RemoveEmptyEntries);

    private static Dictionary<int, Dictionary<char, int>> GetColumnFrequencies()
    {
        var colFreq = new Dictionary<int, Dictionary<char, int>>();

        for (var i = 0; i < Data[0].Length; i++) colFreq[i] = new Dictionary<char, int>();

        foreach (var line in Data)
            for (var i = 0; i < line.Length; i++)
            {
                if (!colFreq[i].ContainsKey(line[i])) colFreq[i][line[i]] = 0;
                colFreq[i][line[i]]++;
            }

        return colFreq;
    }

    private static string SolvePartOne()
    {
        var colFreq = GetColumnFrequencies();
        var result = new char[Data[0].Length];
        for (var i = 0; i < Data[0].Length; i++) result[i] = colFreq[i].OrderByDescending(kvp => kvp.Value).First().Key;

        return new string(result);
    }

    private static string SolvePartTwo()
    {
        var colFreq = GetColumnFrequencies();
        var result = new char[Data[0].Length];
        for (var i = 0; i < Data[0].Length; i++) result[i] = colFreq[i].OrderBy(kvp => kvp.Value).First().Key;

        return new string(result);
    }

    public static void Run()
    {
        Console.WriteLine($"Part One: {SolvePartOne()}");
        Console.WriteLine($"Part Two: {SolvePartTwo()}");
    }
}

File: 2017/06.fs
namespace _2017

open System
open System.Collections.Generic;

module _06 =
    let Data = (Utils.GetInputData 6).Split('\t', StringSplitOptions.RemoveEmptyEntries) |> Array.map int

    let redistribute(banks: int[]) =
        let len = banks.Length
        let maxBlocks = Array.max banks
        let index = Array.findIndex(fun i -> i = maxBlocks) banks
        banks[index] <- 0
        for i in 1 .. maxBlocks do
            banks[(index + i) % len] <- banks[(index + i) % len] + 1

    let solvePartOne() =
        let seenConfigurations = HashSet<string>()
        let rec distribute cycles =
            let config = String.Join(',', Data)
            if seenConfigurations.Contains(config) then cycles
            else
                seenConfigurations.Add(config) |> ignore
                redistribute Data
                distribute (cycles + 1)
        distribute 0

    let solvePartTwo() =
        let seenConfigurations = Dictionary<string, int>()
        let rec distribute (banks: int[]) cycles =
            let config = String.Join(',', banks)
            if seenConfigurations.ContainsKey(config) then
                cycles - seenConfigurations.[config]
            else
                seenConfigurations[config] <- cycles
                redistribute banks
                distribute banks (cycles + 1)
        distribute (Array.copy Data) 0

    let Run() =
        printfn $"Part One: {solvePartOne()}"
        printfn $"Part Two: {solvePartTwo()}"